cmd/consensus.cpp

```diff
@@ -238,8 +239,7 @@ void init_pre_state(const nlohmann::json& pre, StateBuffer& state) {
         Bytes code{from_hex(j["code"].get<std::string>()).value()};
         if (!code.empty()) {
             account.incarnation = 1;
-            ethash::hash256 hash{keccak256(code)};
-            std::memcpy(account.code_hash.bytes, hash.bytes, kHashLength);
+            account.code_hash = bit_cast<evmc::bytes32>(keccak256(code));
```

Comment:
I think in this case a reinterpret_cast is more appropriate. Should save an intermediate copy.
Maybe the compiler takes care of removing memcpy but writing in the code directly makes more clear the intention

Reply:
> I think in this case a reinterpret_cast is more appropriate. Should save an intermediate copy.
> Maybe the compiler takes care of removing memcpy but writing in the code directly makes more clear the intention

`reinterpret_cast` is dangerous, please see the comment in [Chromium's bit_cast](https://chromium.googlesource.com/chromium/src/base/+/refs/heads/main/bit_cast.h#15).

Reply:
Also, I don't agree that `reinterpret_cast` reflects the intention better. The intention is to cast a type that is a wrapper over 32 bytes to another type that is another wrapper of 32 bytes. `reinterpret_cast` is a vague operation with few guarantees from the standard, while `bit_cast`'s purpose is exactly this, cast to a different type but preserve the actual bytes. How `bit_cast` is actually implemented is a separate concern.

Reply:
reinterpret_cast is dangerous when casting different types but in this case we have the very same type with same size and same memory alignment under a different symbol name (always 32 bytes) thus I consider heavier to
* create the destination object
* memcopy data from source
* copy contruct target 
rather than 
* treat source mem location "as if it was of target type"
* copy construct from there

Also in this case the preservation of actual bytes out of scope as the original object is temp and discarded quickly.
Anyway ... I can live with it.

Reply:
On my Linux machine (Ubuntu 20.10, gcc 10.2.0) `evmc::bytes32` (and `evmc_bytes32`) has alignment of 1, while `ethash::hash256` has alignment of 8.

Reply:
Cuz you're considering the alignment of ethash::hash256 **union struct** where the alignment is taken from first type (uint64_t) while evmc::bytes32 though a struct has the alignment of an array of bytes (thus 1)

```
    hash256 h{};
    assert(alignof(h) != alignof(h.bytes);
```

In fact I was reintpret casting hash **bytes** not the whole hash256